documentation/architecture/ADR1:dependency-injection.md

```diff
@@ -0,0 +1,76 @@
+## Context
+##### What is DI?
+
+In DI, client code does not create it's own dependencies.
+```kotlin
+// Without DI, Repository creates it's own PocketEndpoint
+class Repository {
+    val endpoint = PocketEndpoint()
+}
+
+class Parent {
+    val repo = Repository()
+}
+```
+```kotlin
+// With DI, endpoint is injected (passed) through the constructor...
+class Repository(val endpoint: Endpoint) {
+}
+
+class Parent {
+    val repo = Repository(PocketEndpoint())
+}
+
+// ... or through a field
+class Repository {
+    var endpoint: Endpoint
+}
+
+class Parent {
+    val repo = Repository()
+
+    init {
+        repo.endpoint = PocketEndpoint()
+    }
+}
+```
+See https://en.wikipedia.org/wiki/Dependency_injection
+
+##### Why inject?
+
+Injection makes testing simpler. When testing, we don't want our repos to make normal network calls (these make tests flaky, as calls can fail or return unexpected data). So we provide fake endpoints that return the data we need. This is much simpler with DI.
+
+It also makes it easier to follow the adage "program to an interface, not an implementation." Client code (in this case, the Repository) can define the interface upon which they depend, and then never even import the actual implementation. This helps us to separate concerns.
+
+##### Dependency injection vs service locators
+
+We use a service locator pattern. This is similar to, but distinct from dependency injection.
+
+In dependency injection, client code declares that they use an object (either in their constructor, or a field), and it is provided from outside. The client code has no idea how this occurs. In the service locator pattern, client code has a hard dependency upon a service locator, which is used to fulfill dependencies. The above example can be rewritten as a service locator as follows:
+```kotlin
+class Repository(serviceLocator: ServiceLocator) {
+    val endpoint = serviceLocator.endpoint()
+}
+```
+
```

Comment:
Could the below be put in a `Tradeoffs` section?
I'd mention lifecycle management here, as well as any other problems, and how we're choosing to accept those problems/mitigate them.